src/pkt_checker.rs

```diff
@@ -0,0 +1,206 @@
+            let (is_complete, is_read_resp, is_out_of_order, dqpn, end_psn) = {
+                let guard = map
+                    .lock()
+                    .map_err(|_| Error::LockPoisoned("recv packet map lock"))?;
+                (
+                    guard.is_complete(),
+                    guard.is_read_resp(),
+                    guard.is_out_of_order(),
+                    guard.dqpn(),
+                    guard.end_psn(),
+                )
+            };
+            // send ack
+            if is_complete {
+                info!("Complete: {:?}", &msn);
+                if !is_read_resp {
+                    // If we are not in read response, we should send ack
+                    let command =
+                        RespCommand::Acknowledge(RespAckCommand::new_ack(dqpn, msn, end_psn));
+                    self.send_queue
+                        .send(command)
+                        .map_err(|_| Error::PipeBroken("packet checker send queue"))?;
+                } else if let Some(ctx) = self
+                    .read_op_ctx_map
+                    .read()
+                    .map_err(|_| Error::LockPoisoned("read_op_ctx_map lock"))?
+                    .get(&msn)
+                {
+                    if let Err(e) = ctx.set_result(()) {
+                        error!("Set result failed {:?}", e);
+                    }
+                } else {
+                    error!("No read op ctx found for {:?}", msn);
+                }
+                remove_list.push_back(msn);
+            } else if is_out_of_order {
+                // TODO: what should we put in NACK packet?
+                let command = RespCommand::Acknowledge(RespAckCommand::new_nack(
+                    dqpn,
+                    Msn::default(),
+                    end_psn,
+                    Psn::default(),
+                ));
+                self.send_queue
+                    .send(command)
+                    .map_err(|_| Error::PipeBroken("packet checker send queue"))?;
+                panic!("send nack command")
+            } else {
+                // everthing is fine, do nothing
+            }
+        }
+
+        // remove the completed recv_pkt_map
+        if !remove_list.is_empty() {
+            let mut guard = self
+                .recv_pkt_map
+                .write()
+                .map_err(|_| Error::LockPoisoned("recv_pkt_map lock"))?;
+            for dqpn in &remove_list {
+                let _: Option<Arc<Mutex<RecvPktMap>>> = guard.remove(dqpn);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::HashMap,
+        sync::{mpsc, Arc, Mutex, RwLock},
+        thread::sleep,
+        time::Duration,
+    };
+
+    use crate::{
+        op_ctx::ReadOpCtx,
+        recv_pkt_map::RecvPktMap,
+        types::{Msn, Psn, Qpn},
+    };
+
+    use super::PacketChecker;
+
+    #[test]
+    fn test_packet_checker() {
```

Comment:
I recommend also check the error case of the functions above.

Reply:
I have added one. But more needs to be added for other unit tests. I will keep adding.